src/node_api.cc

```diff
@@ -690,18 +727,19 @@ v8::Local<v8::Object> CreateFunctionCallbackData(napi_env env,
   v8::Local<v8::Context> context = isolate->GetCurrentContext();
 
   v8::Local<v8::ObjectTemplate> otpl;
-  ENV_OBJECT_TEMPLATE(env, function_data, otpl, v8impl::kFunctionFieldCount);
+  ENV_OBJECT_TEMPLATE(env, function_data, otpl, v8impl::kInternalFieldCount);
   v8::Local<v8::Object> cbdata = otpl->NewInstance(context).ToLocalChecked();
 
+  CallbackBundle* bundle = new CallbackBundle();
+  bundle->cb[kFunctionIndex] = cb;
+  bundle->cb_data = data;
+  bundle->env = env;
+  bundle->BindLifecycleTo(env->isolate, cbdata);
+
   cbdata->SetInternalField(
-      v8impl::kEnvIndex,
-      v8::External::New(isolate, static_cast<void*>(env)));
-  cbdata->SetInternalField(
-      v8impl::kFunctionIndex,
-      v8::External::New(isolate, reinterpret_cast<void*>(cb)));
-  cbdata->SetInternalField(
-      v8impl::kDataIndex,
-      v8::External::New(isolate, data));
+      v8impl::kCallbackBundleIndex,
+      v8::External::New(isolate, reinterpret_cast<void*>(bundle)));
```

Comment:
We can actually improve performance even more if
```C++
cbdata = v8::External::New(isolate, reinterpret_cast<void*>(bundle))
```
and then we return the `v8::External` directly.

No more need for `SetInternalField()/GetInternalField()`, and we don't need the object templates either.

This could be a further improvement, though that doesn't necessarily need to be done in this PR.